### DISCUSSION TITLE
Utilizing multiple gym environments

### DISCUSSION THREAD

**Discussion by Mgineer117** [PKU-Alignment/safety-gymnasium#66] [Q&A]
Hi!

I am trying to run multiple environments, but I keep getting this error during training: ** World has starting cost! {'cost_out_of_boundary': 1.0, 'cost_sum': 1.0}**.

I have used customized environments with circle task, but only gave a change to the size of circle and distance of Sigwalls. I also have double-checked with no mistake in creating a customized environments. Thus, I am assuming the error arises because I used multiple environments.

Can anyone provide a suggestion how I can fix this error?

The details of error is below:

Traceback (most recent call last):
  File "/home/minjae/Documents/Safety_Gym/main.py", line 119, in <module>
    main_loop()
  File "/home/minjae/Documents/Safety_Gym/main.py", line 116, in main_loop
    algo.train_CPOMeta(writer, save_info_obj)
  File "/home/minjae/Documents/Safety_Gym/algos/CPOMeta.py", line 591, in train_CPOMeta
    memory, log = self.collect_samples(self.envs[batch_iter], self.meta_policy, seed)
  File "/home/minjae/Documents/Safety_Gym/algos/CPOMeta.py", line 298, in collect_samples
    memory, log = collect_trajectory(0, None, env, policy,
  File "/home/minjae/Documents/Safety_Gym/algos/CPOMeta.py", line 41, in collect_trajectory
    state, info = env.reset(seed=seed)
  File "/home/minjae/.local/lib/python3.10/site-packages/gymnasium/wrappers/time_limit.py", line 75, in reset
    return self.env.reset(**kwargs)
  File "/home/minjae/.local/lib/python3.10/site-packages/gymnasium/wrappers/order_enforcing.py", line 61, in reset
    return self.env.reset(**kwargs)
  File "/home/minjae/.local/lib/python3.10/site-packages/gymnasium/wrappers/env_checker.py", line 57, in reset
    return self.env.reset(**kwargs)
  File "/home/minjae/.local/lib/python3.10/site-packages/safety_gymnasium/builder.py", line 184, in reset
    assert cost['cost_sum'] == 0, f'World has starting cost! {cost}'
AssertionError: World has starting cost! {'cost_out_of_boundary': 1.0, 'cost_sum': 1.0}

**Comment by muchvo** [comment]
Can you provide minimal code to explain the change you made for environments?

**Reply by Mgineer117** [reply]
This is example of my stochastic environment which randomly choose value for the radius of circle and distance between Sigwalls.

        def generate_number(min_value, max_value, is_int=False):
            random_number = random.uniform(min_value, max_value)
            if is_int:
                rounded_number = round(random_number)
            else:
                rounded_number = round(random_number, 2)
        
            return rounded_number


--------------------------------
        class StcircleLevel1(StcircleLevel0):
            def __init__(self, config) -> None:
                super().__init__(config=config)
                
                scale = generate_number(0.45, 0.85)
                distance = round(self.circle.radius * scale, 3)
                num_pillars = generate_number(3, 6, is_int=True)
                num_hazards = generate_number(4, 6, is_int=True)

                self._add_geoms(Sigwalls(num = 2, locate_factor=distance, size=3.5, is_constrained=True))
                self._add_geoms(Pillars(num=num_pillars, size=0.1, height=0.2, is_constrained=False))
                self._add_geoms(Hazards(num=num_hazards, size=0.12, keepout=0.18, is_constrained=False))


---------------------------------------------------
The bug vanishes when I rerun the training, but somehow again pops up in some cases.

**Comment by muchvo** [comment]
```
class CircleLevel0(BaseTask):
    """An agent want to loop around the boundary of circle."""

    def __init__(self, config) -> None:
        super().__init__(config=config)

        self.num_steps = 500

        self.agent.placements = [(-0.8, -0.8, 0.8, 0.8)]
        self.agent.keepout = 0

        self.lidar_conf.max_dist = 6

        # Reward for circle goal (complicated formula depending on pos and vel)
        self.reward_factor: float = 1e-1

        self._add_geoms(Circle())
```
You should notice that the agent will randomly be placed in points which are constrained in `self.agent.placements = [(-0.8, -0.8, 0.8, 0.8)]`.  Then, If the agent is placed outside Sigwalls at the beginning. You will encounter assert error.

**Reply by Mgineer117** [reply]
It now makes sense. I appreciate that.
